pkg/policy/existing.go

```diff
@@ -29,36 +29,23 @@ func (pc *PolicyController) processExistingResources(policy *kyverno.ClusterPoli
 			continue
 		}
 
-		for _, kind := range rule.MatchResources.Kinds {
-			k := common.GetFormatedKind(kind)
-			logger = logger.WithValues("rule", rule.Name, "kind", k)
-			namespaced, err := pc.rm.GetScope(k)
-			if err != nil {
-				if err := pc.registerResource(k); err != nil {
-					logger.Error(err, "failed to find resource", "kind", k)
-					continue
-				}
-
-				namespaced, _ = pc.rm.GetScope(k)
-			}
-
-			// this tracker would help to ensure that even for multiple namespaces, duplicate metric are not generated
-			metricRegisteredTracker := false
-
-			if !namespaced {
-				pc.applyAndReportPerNamespace(policy, k, "", rule, logger.WithValues("kind", k), &metricRegisteredTracker)
-				continue
-			}
+		match := rule.MatchResources
+		exclude := rule.ExcludeResources
 
-			namespaces := pc.getNamespacesForRule(&rule, logger.WithValues("kind", k))
-			for _, ns := range namespaces {
-				// for kind: Policy, consider only the namespace which the policy belongs to.
-				// for kind: ClusterPolicy, consider all the namespaces.
-				if policy.Namespace == ns || policy.Namespace == "" {
-					pc.applyAndReportPerNamespace(policy, k, ns, rule, logger.WithValues("kind", k).WithValues("ns", ns), &metricRegisteredTracker)
-				}
-			}
+		for _, value := range match.Any {
+			pc.processExistingKinds(value.ResourceDescription.Kinds, policy, rule, logger)
 		}
+		for _, value := range match.All {
+			pc.processExistingKinds(value.ResourceDescription.Kinds, policy, rule, logger)
+		}
+		for _, value := range exclude.All {
+			pc.processExistingKinds(value.ResourceDescription.Kinds, policy, rule, logger)
+		}
+		for _, value := range exclude.Any {
+			pc.processExistingKinds(value.ResourceDescription.Kinds, policy, rule, logger)
+		}
```

Comment:
`processExistingKinds` takes the matching kind to fetch the list of resources. Why do we pass exclude kinds list here?

We should update this logic to filter out resources excluded from any/all:
https://github.com/kyverno/kyverno/blob/e6317da597e2c0b18e9406feb4bd0a9309612ba2/pkg/policy/common.go#L223

Note that we need to exclude each entry individually in `any`, but all (merged) entries in `all`. For example:
The following rule excludes all Pods and all resources in namespace staging.
```yaml
exclude:
  any:
  - resources:
      kinds:
      - Pod
  - resources:
      namespaces:
      - staging
```
While this rule excludes the Pod from the staging namespace only.
```yaml
exclude:
  all:
  - resources:
      kinds:
      - Pod
  - resources:
      namespaces:
      - staging
```